### PR TITLE
UI improvements for the Copilot rename feature

### PR DESCRIPTION
**Pull request by AmadeusW**
This PR makes UI changes to the Copilot powered inline rename feature

Pictured below: Keyboard shortcut or button mouse click triggers rename and, if applicable (because of Feature Flag), records that from now on suggestions should be retrieved automatically. Next time rename UI opens, suggestions are obtained automatically.

smart rename ui refresh

Specifically, this PR brings the following changes:

- Use the results panel (instead of the dropdown) in both automatic and explicit modes
  - explicit mode using result panel: 
image

- Mouse double click to commit rename (from user feedback)
- Fix issue where selecting rename suggestions may change width of the popup
  -  before:
smart rename ui resize bug
  - after: 
smart rename ui resize fixed

- Fix strings (hint and tooltip) that were swapped for automatic and explicit modes
  - image
  - image

- Add delay prior to making the request
  - demo of delay prior to integrating with #73949 :
smart rename ui delay without cancelation

- The button gets purple highlight only to designate its toggled state, i.e. rename is in automatic mode (through feature flag), and user opted into automatic suggestions (through button click or Ctrl+Space)
  - image
  - In light mode, it's a blue border: image

### CODE REVIEW

File: src/Deployment/Properties/launchSettings.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "profiles": {
     "Visual Studio Extension": {
       "commandName": "Executable",
-      "executablePath": "$(DevEnvDir)devenv.exe",
+      "executablePath": "C:\\Program Files\\Microsoft Visual Studio\\2022\\main\\Common7\\IDE\\devenv.exe",
```
Comment (CyrusNajmabadi) [inline src/Deployment/Properties/launchSettings.json:5]: revert please.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 2]
```diff
@@ -205,12 +205,19 @@ private void Adornment_KeyDown(object sender, KeyEventArgs e)
...
                         if (_viewModel.SmartRenameViewModel is not null && _viewModel.SmartRenameViewModel.GetSuggestionsCommand.CanExecute(null))
                         {
-                            // If smart rename can use automatic behavior, enable it
-                            if (!_viewModel.SmartRenameViewModel.IsUsingDropdown)
+                            if (_viewModel.SmartRenameViewModel.SupportsAutomaticSuggestions)
                             {
-                                _viewModel.SmartRenameViewModel.IsSuggestionsPanelCollapsed = !_viewModel.SmartRenameViewModel.IsSuggestionsPanelCollapsed;
+                                // If smart rename can use automatic behavior, use the button to toggle it.
+                                _viewModel.SmartRenameViewModel.ToggleAutomaticSuggestions();
+                                if (_viewModel.SmartRenameViewModel.IsAutomaticSuggestionsEnabled)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:212]: i really don't understand the relationship between SupportsAutomaticSuggestions, ToggleAutomaticSuggestions and IsAutomaticSuggestionsEnabled.  It sounds like you're saying "if we support automatic suggestions, switch to the opposite of whatever our current setting is.  I don't get that part.  Wouldn't this cause us to just switch on/off/on/off each time someone opens the dialog?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 3]
```diff
@@ -205,12 +205,19 @@ private void Adornment_KeyDown(object sender, KeyEventArgs e)
...
+                                {
+                                    _viewModel.SmartRenameViewModel.GetSuggestionsCommand.Execute(null);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:214]: named arg for null.  i have no idea what it means :)

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs [hunk 4]
```diff
@@ -276,7 +283,7 @@ private void RenameUserInput_PreviewKeyDown(object sender, KeyEventArgs e)
                     {
                         _viewModel.IdentifierText = newIdentifier;
                         // Place the cursor at the end of the input text box.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/Adornment/RenameFlyout.xaml.cs:285]: this comment is wrong.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml [hunk 5]
```diff
@@ -319,6 +254,7 @@
             <Setter Property="Background" Value="Transparent" />
             <Setter Property="Template" Value="{StaticResource ComboBoxItemTemplate}" />
             <Setter Property="Margin" Value="2,0,2,0" />
+            <Setter Property="Padding" Value="4,1,24,1" /> <!-- Add 20px to the right over the default template to prevent popup from changing size -->
         </Style>
     </ComboBox.ItemContainerStyle>
 </ComboBox>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml:260]: skipping all the xaml.  maybe @ryzngard  can look at this part.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs [hunk 6]
```diff
@@ -95,19 +93,23 @@ void IRenameUserInput.Focus()
...
-        base.OnApplyTemplate();
+        // Handle the event to avoid stack overflow through passing execution back RenameFlyout.Adornment_GotKeyboardFocus
+        e.Handled = true;
+    }
 
-        _dropDownPopup = (Popup)GetTemplateChild(DropDownPopup)!;
+    private void SuggestionsPanelScrollViewer_MouseDoubleClick(object sender, MouseEventArgs e)
+    {
+        _baseViewModel.Submit();
     }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:105]: does thsi ui work without a mouse?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs [hunk 7]
```diff
@@ -95,19 +93,23 @@ void IRenameUserInput.Focus()
...
 
     private void GetSuggestionsButtonClick(object sender, RoutedEventArgs e)
     {
-        if (_smartRenameViewModel.IsUsingResultPanel)
+        if (_smartRenameViewModel.SupportsAutomaticSuggestions)
         {
-            _smartRenameViewModel.IsSuggestionsPanelCollapsed = !_smartRenameViewModel.IsSuggestionsPanelCollapsed;
-            if (_smartRenameViewModel.IsSuggestionsPanelExpanded)
+            _smartRenameViewModel.ToggleAutomaticSuggestions();
+            if (_smartRenameViewModel.IsAutomaticSuggestionsEnabled)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameUserInputComboBox.xaml.cs:112]: ame point about not understanding this pattern.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 8]
```diff
@@ -48,8 +49,22 @@ internal sealed partial class SmartRenameViewModel : INotifyPropertyChanged, IDi
...
+    public bool SupportsAutomaticSuggestions { get; }
+
+    public bool IsAutomaticSuggestionsEnabled { get; private set; }
+
+    public void ToggleAutomaticSuggestions()
+    {
+        if (!SupportsAutomaticSuggestions)
+        {
+            return;
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:62]: blank line after }.  I'm surprised you didn't get a warning.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 9]
```diff
@@ -48,8 +49,22 @@ internal sealed partial class SmartRenameViewModel : INotifyPropertyChanged, IDi
     public string StatusMessage => _smartRenameSession.StatusMessage;
 
     public bool StatusMessageVisibility => _smartRenameSession.StatusMessageVisibility;
-    public bool IsUsingResultPanel { get; set; }
-    public bool IsUsingDropdown { get; set; }
+
+    public bool SupportsAutomaticSuggestions { get; }
+
+    public bool IsAutomaticSuggestionsEnabled { get; private set; }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:55]: doc what thread we can call these on.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 10]
```diff
@@ -48,8 +49,22 @@ internal sealed partial class SmartRenameViewModel : INotifyPropertyChanged, IDi
...
+
+    public void ToggleAutomaticSuggestions()
+    {
+        if (!SupportsAutomaticSuggestions)
+        {
+            return;
+        }
+        IsAutomaticSuggestionsEnabled = !IsAutomaticSuggestionsEnabled;
+        // Use existing option (true if user does not wish to get suggestions automatically) to honor user's choice from before the refactoring.
+        _globalOptionService.SetGlobalOption(InlineRenameUIOptionsStorage.CollapseSuggestionsPanel, !IsAutomaticSuggestionsEnabled);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:65]: i'm so confused.  we flipped hte option already in teh line above.  and now we're flipping this back?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 11]
```diff
@@ -48,8 +49,22 @@ internal sealed partial class SmartRenameViewModel : INotifyPropertyChanged, IDi
...
-    public bool IsUsingResultPanel { get; set; }
-    public bool IsUsingDropdown { get; set; }
+
+    public bool SupportsAutomaticSuggestions { get; }
+
+    public bool IsAutomaticSuggestionsEnabled { get; private set; }
+
+    public void ToggleAutomaticSuggestions()
+    {
+        if (!SupportsAutomaticSuggestions)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:59]: if Toggling already checks if we support this, then why do the callers *also* check this?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 12]
```diff
@@ -70,34 +85,19 @@ internal sealed partial class SmartRenameViewModel : INotifyPropertyChanged, IDi
...
-    }
+    public bool IsSuggestionsPanelExpanded => HasSuggestions;
 
     public string GetSuggestionsTooltip
-        => IsUsingDropdown
-        ? EditorFeaturesWpfResources.Get_AI_suggestions
-        : EditorFeaturesWpfResources.Toggle_AI_suggestions;
+        => SupportsAutomaticSuggestions
+        ? EditorFeaturesWpfResources.Toggle_AI_suggestions
+        : EditorFeaturesWpfResources.Get_AI_suggestions;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:95]:
```suggestion
            ? EditorFeaturesWpfResources.Toggle_AI_suggestions
            : EditorFeaturesWpfResources.Get_AI_suggestions;
```

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 13]
```diff
@@ -124,20 +124,20 @@ public string SubmitTextOverride
...
 
-        var getSuggestionsAutomatically = _globalOptionService.GetOption(InlineRenameUIOptionsStorage.GetSuggestionsAutomatically);
-        IsUsingResultPanel = getSuggestionsAutomatically;
-        IsUsingDropdown = !IsUsingResultPanel;
         SetupTelemetry();
-        if (IsUsingResultPanel && IsSuggestionsPanelExpanded)
+
+        this.SupportsAutomaticSuggestions = !_globalOptionService.GetOption(InlineRenameUIOptionsStorage.GetSuggestionsAutomatically);
+        this.IsAutomaticSuggestionsEnabled = this.SupportsAutomaticSuggestions && !_globalOptionService.GetOption(InlineRenameUIOptionsStorage.CollapseSuggestionsPanel);
+        if (this.SupportsAutomaticSuggestions && this.IsAutomaticSuggestionsEnabled)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:131]: why check both.  can't IsAutomaticSuggestionsEnabled say that it's not enabled if we don't support automatic suggestions?  There is a ton of boolean values floating around that need to be checked in odd ways.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 14]
```diff
@@ -124,20 +124,20 @@ public string SubmitTextOverride
 
         GetSuggestionsCommand = new DelegateCommand(OnGetSuggestionsCommandExecute, null, threadingContext.JoinableTaskFactory);
 
-        var getSuggestionsAutomatically = _globalOptionService.GetOption(InlineRenameUIOptionsStorage.GetSuggestionsAutomatically);
-        IsUsingResultPanel = getSuggestionsAutomatically;
-        IsUsingDropdown = !IsUsingResultPanel;
         SetupTelemetry();
-        if (IsUsingResultPanel && IsSuggestionsPanelExpanded)
+
+        this.SupportsAutomaticSuggestions = !_globalOptionService.GetOption(InlineRenameUIOptionsStorage.GetSuggestionsAutomatically);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:129]: i don't understand the `!` here.  The negation of things is *really* confusing.  If hte option is "GetSuggestionsAutomatically" and the user has that set to true... why are we setting SupportsAutomaticSuggestions to false.  This is very confusing.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 15]
```diff
@@ -146,14 +146,24 @@ private void OnGetSuggestionsCommandExecute()
...
+            var isAutomatic = parameter is bool b && b;
+            _getSuggestionsTask = GetSuggestionsTaskAsync(isAutomatic, _cancellationTokenSource.Token).CompletesAsyncOperation(listenerToken);
+        }
+    }
+
+    private async Task GetSuggestionsTaskAsync(bool isAutomatic, CancellationToken cancellationToken)
+    {
+        if (isAutomatic)
+        {
+            await Task.Delay(_smartRenameSession.AutomaticFetchDelay, cancellationToken).ConfigureAwait(true);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:160]: doc why ConfigureAwait(true).

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 16]
```diff
@@ -146,14 +146,24 @@ private void OnGetSuggestionsCommandExecute()
...
+        }
+        if (cancellationToken.IsCancellationRequested)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:162]: need to deal with the UI possibly being gone at this point as well.

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 17]
```diff
@@ -146,14 +146,24 @@ private void OnGetSuggestionsCommandExecute()
...
+        {
+            return;
         }
+        _ = await _smartRenameSession.GetSuggestionsAsync(_cancellationTokenSource.Token);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:166]: this is super confusing.  We have a cancellation token passed in *and* an ambient cancellation token source?  why is this the right token.  this needs docs.

also... why are we ignoring the result of this call?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 18]
```diff
@@ -165,20 +175,17 @@ private void SessionPropertyChanged(object sender, PropertyChangedEventArgs e)
...
-            foreach (var name in _smartRenameSession.SuggestedNames)
+            // Set limit of 3 results
+            foreach (var name in _smartRenameSession.SuggestedNames.Take(3))
             {
-                if (++count > 3 && IsUsingResultPanel)
-                {
-                    // Set limit of 3 results when using the result panel
-                    break;
-                }
                 SuggestedNames.Add(name);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:181]: can you just do `.AddRAnge(...Take(3));`?

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 19]
```diff
@@ -165,20 +175,17 @@ private void SessionPropertyChanged(object sender, PropertyChangedEventArgs e)
...
             }
 
             // Changing the list may have changed the text in the text box. We need to restore it.
             BaseViewModel.IdentifierText = textInputBackup;
 
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsSuggestionsPanelCollapsed)));
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsSuggestionsPanelExpanded)));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:188]: wait... why are we invoking both of these?

**Comment by CyrusNajmabadi** [comment]
Please don't force-push reviews.  It makes seeing what changed impossible.  Note: when you finally submit you can squash if you want.

**Comment by ryzngard** [comment]
For UI changes please make sure the following happens: 

1. If you added new items make sure to test with [accessibility insights](https://accessibilityinsights.io/)
2. Please include a screenshot/gif of the new UI

**Comment by AmadeusW** [comment]
> Please don't force-push reviews. It makes seeing what changed impossible. Note: when you finally submit you can squash if you want.

Sorry, I rebased to integrate with another PR

**Comment by AmadeusW** [comment]
> For UI changes please make sure the following happens:
> 
> 1. If you added new items make sure to test with [accessibility insights](https://accessibilityinsights.io/)
> 2. Please include a screenshot/gif of the new UI

1. No new items added but I'll do a pass with accinsights
2. Updated PR with gif and screenshots for itemized items

**Comment by CyrusNajmabadi** [comment]
Merge commits still work fine for that :-)

File: src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs [hunk 20]
```diff
@@ -220,12 +224,39 @@ public void Commit(string finalIdentifierName)
...
+    /// <summary>
+    /// When smart rename operates in explicit mode, this method gets the suggestions.
+    /// When smart rename operates in automatic mode, this method toggles the automatic suggestions, 
+    /// and gets the suggestions if it was just enabled.
+    /// </summary>
+    public void ToggleOrTriggerSuggestions()
+    {
+        if (this.SupportsAutomaticSuggestions)
+        {
+            this.IsAutomaticSuggestionsEnabled = !this.IsAutomaticSuggestionsEnabled;
```
Comment (Cosifne) [inline src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs:244]: 🤔Could you explain about this? why it needs to be flipped?

**Comment by Cosifne** [comment]
Also, target release/dev17.11 if you want to make it into 17.11.
cc @arkalyanms this would be an M2 change

**Comment by AmadeusW** [comment]
done, thank you!

**Comment by Cosifne** [comment]
Ah I retarget the PR. Seems like the base is based on the main branch after the snap. So we need to manually pick change into 17.11

**Review by Cosifne — APPROVED** [review]

**Pull request merged**